p2p/src/simulated/network.rs

```diff
@@ -482,6 +520,151 @@ impl crate::Receiver for Receiver {
+                                    }
+                                };
+                                if let Err(err) = mailbox.send((dialer.clone(), message)).await {
+                                    error!(?err, "failed to send message to mailbox");
+                                    break;
+                                }
+                            }
+                        }
+                    });
+                }
+            }
+        });
+
+        result
+    }
+
+    fn register(&mut self, channel: Channel) -> Result<mpsc::UnboundedReceiver<Message>, Error> {
+        let mut mailboxes = self.mailboxes.lock().unwrap();
+
+        // Error if channel already exists
+        if mailboxes.contains_key(&channel) {
+            return Err(Error::ChannelAlreadyRegistered(channel));
+        }
+
+        // Insert new channel
+        let (sender, receiver) = mpsc::unbounded();
+        mailboxes.insert(channel, sender);
+        Ok(receiver)
+    }
+}
+
+// A unidirectional link between two peers.
+// Messages can be sent over the link with a given latency, jitter, and success rate.
+#[derive(Clone)]
+struct Link {
+    sampler: Normal<f64>,
+    success_rate: f64,
+    inbox: mpsc::UnboundedSender<(Channel, Bytes)>,
+}
+
+impl Link {
+    fn new<E: Spawner + RNetwork<Listener, Sink, Stream>>(
+        runtime: &mut E,
+        dialer: PublicKey,
+        socket: SocketAddr,
+        sampler: Normal<f64>,
+        success_rate: f64,
+        max_size: usize,
+    ) -> Self {
+        let (inbox, mut outbox) = mpsc::unbounded();
+        let result = Self {
+            sampler,
+            success_rate,
+            inbox,
+        };
+
+        // Spawn a task that will wait for messages to be sent to the link and then send them
+        // over the network.
+        runtime.spawn("sender", {
+            let runtime = runtime.clone();
+            async move {
+                // Dial the peer and handshake by sending it the dialer's public key
+                let (mut sink, _) = runtime.dial(socket).await.unwrap();
+                if let Err(err) = send_frame(&mut sink, &dialer, max_size).await {
+                    error!(?err, "failed to send public key to dialee");
+                    return;
+                }
+
+                // For any item placed in the inbox, send it to the sink
+                while let Some((channel, message)) = outbox.next().await {
+                    let mut data = bytes::BytesMut::with_capacity(SIZE_OF_CHANNEL + message.len());
+                    data.extend_from_slice(&channel.to_be_bytes());
+                    data.extend_from_slice(&message);
+                    let data = data.freeze();
+                    send_frame(&mut sink, &data, max_size).await.unwrap();
+                }
+            }
+        });
+
+        result
+    }
+
+    // Send a message over the link.
+    async fn send(&self, channel: Channel, message: Bytes) -> Result<(), Error> {
```

Comment:
Is it impossible to turn this into `&mut self`? Then we could avoid cloning for each message.

Reply:
It looks like you already clone it above (in `handle_task`), so you should have an exclusive reference to it.

Right now you are double-cloning (once from `self.links` and then here before you use it).